### PR TITLE
Add code for submission

### PR DESCRIPTION
**Pull request by jbcdnr**
If you have any comments/concerns on how we should handle submissions, let's discuss it here!

### Our code is now a python package

```python
from auto_train import Task, Batch
```

### Docker for submission

I created a base Dockerfile with PyTorch and our dependencies. There is a sample submission folder to let them quickly start. I pushed our base docker image to [Dockerhub](https://cloud.docker.com/u/jbcdnr/repository/docker/jbcdnr/auto-train) so they simply have to build on top of it to install their dependencies (only public packages/links).

## Next

- [x] Add a LaTeX template for the report
- [x] Move docker image under organisation name ?

### CODE REVIEW

**Comment by tvogels** [comment]
Nice! I think it is good that our stuff is in a package. Docker stuff looks good, too.
I wonder, however, if the starter kit is the right place for our library + docker image. It clutters the repository a bit, I feel. I think it would be nice if this repo only has the a sample submission example. 

We do, however, need people to be able to see what the docker environment is like, and also see the example task.py (maybe that should be called auto_train.py now)

Maybe we can create a directory `environment` and move `Dockerfile`, `setup.py` and `src` into there.

Docker note: we can probably remove a whole bunch of those dependencies.

Let's have a quick chat tomorrow. I can have a look into your proposed next steps as well.

**Comment by martinjaggi** [comment]
if you want to make clear it's independent, better call the folder `sample-submission` than `environment` ?

**Comment by tvogels** [comment]
`environment` would be anything that is *not* part of the submission

I would propose this structure:
~~~
sample-submission/
    Dockerfile
    README.md
    train.py
    report/
        template.tex
environment/
    Dockerfile
    setup.py
    src/
        task.py
README.md
~~~

**Comment by tvogels** [comment]
@martinjaggi, do we already have an organization account for Dockerhub? Should I create one? Should it be MLO or a special one for autoTrain?

Edit: I created this:
https://cloud.docker.com/u/autotrain/repository/docker/autotrain/autotrain-env
@jbcdnr and I are currently listed as collaborators, so we should have push rights.

**Comment by martinjaggi** [comment]
looks good

**Comment by tvogels** [comment]
@jbcdnr and I had a quick discussion offline. A few points of discussion:

- We think using `autotrain/autotrain-env` makes sense as a Docker package name. The pull request is updated to use this. @martinjaggi, would you prefer to use the MLBench organization?

- About the TeX report: What should we ask for? Currently we say 4 pages 2-columns. Should there be a clear spec? Should a report be optional? On the one hand, people who submit might want to write something about their submission anyway, on the other hand, we don't really intend to read it anyway.

- The current two-column report template I added is a bit ugly. Anyone knows a better one?

- We currently specify that the submission must be in 1 file, `train.py`, and contestants can install any publicly available dependencies via Docker. The idea behind this is that it should encourage easily adoptable and isolated code. **Q:** Should we put a restriction on the licenses for external dependencies? If they want to use some big internal library, they will have to open-source it first. If they want to use a high performance Fortran library, they will have to open-source python bindings first. The community would benefit from this.

**Comment by martinjaggi** [comment]
docker org: anything is fine

potential closed source dependencies: agreed open source should be highly preferred (and a must for the algorithm part itself), but let's discuss some more where to draw the line

**Comment by tvogels** [comment]
Thanks for the reorganization, @jbcdnr! Some comments:

I would call the `resnet_task` something like `dummy_task` or `example_task` to avoid giving contestants the impression that they can choose their tasks. I don't think we should add more than one task to this repo.

Currently, the abstract class `Task` is quite heavy and the implementation `ResnetTask` is quite light. While this is good for avoiding code duplication, it does prescribe fairly rigid pattern in how the task is constructed (It also prescribes the loss function currently). I would suggest the abstract `Task` to be a mere interface and put all the implementation (at the cost of some duplication) in the tasks themselves. What do you think? Should I make a proposal?

**Comment by jbcdnr** [comment]
I think that when we'll have 5+ tasks to maintain, it will be painful to update common code. As we settled for supervised tasks, the current structure has all the flexibility we need. I propose to keep it like this and update it when we add more tasks if more flexibility is needed, but I am open to discussion.

- I add `_create_criterion`,
- Agreed for name change.

Can I merge this PR to make things appear on master? We will open a new one to continue the work! ;)

**Comment by tvogels** [comment]
As we discussed offline, "In the end we will probably have `Task` just the interface, `SupervisedTask` all the reused code, `Task1`, `Task2` ... 

I committed a proposal for the `Task` interface and an `ExampleTask`. Let's postpone code sharing between tasks like in your proposal until we have more tasks.

**Comment by tvogels** [comment]
Let's merge this and discuss further on Monday.
